src/lib/sql/sql_translator.cpp

```diff
@@ -403,15 +410,103 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_join(const hsql::Join
-
+  auto predicate_condition = translate_operator_type_to_predicate_condition(join_condition->opType);
   auto join_node = JoinNode::make(join_mode, column_references, predicate_condition);
   join_node->set_left_input(left_node);
   join_node->set_right_input(right_node);
-
+  _apply_predicate_expressions(join_node, LQPInputSide::Left, left);
+  _apply_predicate_expressions(join_node, LQPInputSide::Right, right);
   return join_node;
 }
 
+LQPInputSide SQLTranslator::_get_side(const std::shared_ptr<AbstractLQPNode>& left_node,
+                                      const std::shared_ptr<AbstractLQPNode>& right_node, const hsql::Expr* expr) {
+  DebugAssert(expr && expr->type == hsql::kExprColumnRef, "only applicable to columnar expressions");
+  const auto qualified = HSQLExprTranslator::to_qualified_column_name(*expr);
+  if (left_node->find_column(qualified)) {
+    return LQPInputSide::Left;
+  }
+
+  if (right_node->find_column(qualified)) {
+    return LQPInputSide::Right;
+  }
+
+  Fail("Column not found: " + qualified.as_string());
+}
+
+void SQLTranslator::_get_sides_from_operator_expressions(std::vector<const hsql::Expr*> operators,
+                                                         std::function<LQPInputSide(const hsql::Expr*)> get_side,
+                                                         std::vector<const hsql::Expr*>& both,
+                                                         std::vector<const hsql::Expr*>& left,
+                                                         std::vector<const hsql::Expr*>& right) {
+  for (const auto* op : operators) {
+    const bool left_is_column = op->expr && op->expr->type == hsql::kExprColumnRef;
+    const bool right_is_column = op->expr2 && op->expr2->type == hsql::kExprColumnRef;
+
+    if (left_is_column && right_is_column) {
+      auto left_column_side = get_side(op->expr);
+      auto right_column_side = get_side(op->expr2);
+      if (left_column_side == right_column_side) {
+        if (left_column_side == LQPInputSide::Left) {
+          left.push_back(op);
+        } else {
+          right.push_back(op);
+        }
+      } else {
+        both.push_back(op);
+      }
+    } else {
+      if (left_is_column) {
+        if (get_side(op->expr) == LQPInputSide::Left) {
+          left.push_back(op);
+        } else {
+          right.push_back(op);
+        }
+      } else if (right_is_column) {
+        if (get_side(op->expr2) == LQPInputSide::Left) {
+          left.push_back(op);
+        } else {
+          right.push_back(op);
+        }
+      }
+    }
+  }
+}
+
+void SQLTranslator::_apply_predicate_expressions(const std::shared_ptr<AbstractLQPNode>& node, const LQPInputSide side,
+                                                 const std::vector<const hsql::Expr*>& conditions) {
+  if (conditions.empty()) {
+    return;
+  }
+
+  std::vector<std::shared_ptr<AbstractLQPNode>> predicates;
```

Comment:
Why do you have to build a vector here? Can't you create the predicate nodes in the for loop without an auxiliary structure?

Reply:
Technically, yes. 
I really like about the current proposal how creating the LQP nodes and wiring them up are separate tasks, to be done and understood individually. It's easy to see and introspect all created predicates.
To my understanding the alternative would be to start with a lone pointer to keep track of the previous node and then mutate and link on the previous node in one pass.
If you are concerned with stack space or re-allocations, I suggest to adjust the vector to the exact number of elements needed.

Reply:
I pointed out concerns with the same piece of code here: https://github.com/hyrise/hyrise/pull/728#discussion_r172883909 just so this doesn't get lost

Reply:
I don‘t worry about the cost of that structure, I just wonder if it isn‘t more natural to just add the nodes one by one. Apparently it isn‘t more natural for everyone, and I can live with the current state.